### ISSUE TITLE
API Proposal for Classification Feature

### ISSUE THREAD

**Issue by etiennedi** [semi-technologies/weaviate#977]
API for #948 

Three API use cases need to be addressed:

1. Start a classification, see status of ongoing classification
2. View classified items
3. View meta data about a thing which was part of a classification

Here are proposals for all three:

## 1. Trigger Classification, see Status
```
POST /v1/classifications 
{
  "class": "Company",
  "classifyProperties": ["materialGroupOne", "materialGroupTwo"],
  "basedOnProperties": ["description"],
  "type": "knn", // optional, can default to knn
  "k": 3 // optional, can default to something reasonable
}

returns GET response + Header 'Location: /v1/classifications/<id>'
```

```
GET /v1/classifications/<classificationID>
{
  "id": "abcdefgh",
  "class": "Company",
  "classifyProperties": ["materialGroupOne", "materialGroupTwo"],
  "basedOnProperties": ["description"],
  "status": "running",
  "started": <timestamp>,
  "completed": <timestamp> || null,
  "type": "knn"
  "k": 3 
}
```

## 2. View classified items

No new APIs required, as classifications are just cross-refs. Any existing measures to view crossrefs (resolved in GQL `Get` or as beacons in REST `GET /v1/things/<id>` can be used.

### Regular `GET /v1/thing/<id>` without meta parameter
Unchanged, same as always: 
```
GET /v1/things/<id>
{
  "class": "Company",
  "id": "<id>",
  "schema": {
    "description": "Foo bar baz bazzar",
    "materialGroupOne": [
      {
         "beacon": "..."
      }
    ],
   "materialGroupTwo": [
      {
         "beacon": "..."
      }
    ]
  }
}
```

### Get with meta parameter `GET /v1/thing/<id>?meta=true`
```
GET /v1/things/<id>
{
  "class": "Company",
  "id": "<id>",
  "schema": {
    "description": "Foo bar baz bazzar",
    "materialGroupOne": [
      {
         "beacon": "...",
         "meta": null  <----- indicates this has not been classified, but set by a user
      }
    ],
   "materialGroupTwo": [
      {
         "beacon": "...",
         "meta": {
            "classification": {
              "distanceWinning": 0.5,
              "distanceLosing": 0.6,
            },
          },
      }
    ]
  }
  "meta": {
    "classification": {
      "id": "<classificatinId>",
      "completed": <timestamp>,
      "classifiedFields": ["matericalGroupTwo"],
      "scope": ["materialGroupOne", "materialGroupTwo"]  <----- note that the scope was to classify both fields, but since the user had already set materialGroupOne the actual classified Fields (see one row up) is only a single field (materialGroupTwo)
      "basedOn": ["description"],
    }
  }
}
```

**Comment by fefi42** [comment]
What kind of distance metric is used?
I would like to have some kind of (average) distance in the meta object:
```
"meta": {
    "classification": {
      "id": "<classificatinId>",
      "completed": <timestamp>,
      "classifiedFields": ["matericalGroupTwo"],
      "scope": ["materialGroupOne", "materialGroupTwo"]  <----- note that the scope was to classify both fields, but since the user had already set materialGroupOne the actual classified Fields (see one row up) is only a single field (materialGroupTwo)
      "basedOn": ["description"]
      "distance": 0.334
    }
  }
```
This can serve as some kind of confidence or quality measure.

Typo in example this should mean: `"source": "classification"  <----- indicates this has been classified, but set by a user`

**Comment by etiennedi** [comment]
Great point abut the distance, @fefi42. I agree.

What would be the best way to calculate the distance?
- Mean distance between classified item and all k neighbors?

In addition would it be helpful to include how many neighbors were of the same class. 

For example if k=7,
- if 4 point to A and 3 point to B, we classify as A
- if all 7 point to A, we also classify as A

Could we argue that the second classification was done with a higher confidence? If so how would we express this? As part of the distance or as a separate metric?

**Comment by bobvanluijt** [comment]
One additional idea comes to mind.

Being able to set a where filter:

```
POST /v1/classifications 
{
  "class": "Company",
  "classifyProperties": ["materialGroupOne", "materialGroupTwo"],
  "basedOnProperties": ["description"],
  "type": "knn", // optional, can default to knn
  "k": 3 // optional, can default to something reasonable
  "where": {
    // same as where in GQL query
  }
}
```

It would allow users to: ”e.g., only classify companies with more than 1000 employees”

**Comment by bobvanluijt** [comment]
Also, [this](https://github.com/semi-technologies/weaviate/issues/952) feature seems to become more and more important.

`basedOnProperties` should not be set to none-indexable.

**Comment by etiennedi** [comment]
@bobvanluijt 

Great point about the where filter, I also really like the fact that we are enriching the REST API with something that was previously exclusive to the GQL. (Once we have the where filter in one place it should be trivial to also copy it to other places like `GET /v1/things`)

> `basedOnProperties` should not be set to a nonindexable property.

Fully agreed. In fact it's a technical necessity for the field to be indexed, so that we can even do a classification. I will see that I can prioritize #952 right after the first usable version of classifications.

**Comment by bobvanluijt** [comment]
Thanks

An example of a non-indexable use case might be articles. We want the c11y to index the abstract, but not the meta data.

**Comment by fefi42** [comment]
I am currently not sure whats the best way to calculate an average distance, maybe it would be best to just give three values. `Min, arithmetic average, Max` This would help to quickly spot if `n` is chosen correctly. E.g. if `min` is very small but `max` and `mean` are very high this would indicate that `n` is too high.

**Comment by fefi42** [comment]
Wouldn't a where clause also improve performance? Because it reduces the search space.

**Comment by etiennedi** [comment]
> Wouldn't a where clause also (potentially) improve performance? Because it reduces the search space.

Your point is a good one, it highlights that there are two spaces we can limit. Let's take the examples that I have `Product ---MaterialGroup---> Material` and I want to classify products as belonging to a material group. I know how two potential filtering options:

A.) "Only classify those Products which have price > 100"
B.) "Only consider target MaterialGroups where organic==true"

@bobvanluijt which one did you mean when suggesting the where filter? I implicitly assumed we mean A., but the point Felix raises would also apply to B.

**Comment by bobvanluijt** [comment]
> I am currently not sure whats the best way to calculate an average distance

@fefi42 I'm not sure I understand. We want to solve this for the end user in Weaviate rather than have the end user do it, right?

**Comment by fefi42** [comment]
> I am currently not sure whats the best way to calculate an average distance
Yes we show this to the enduser. I would just start with the normal average, nothing fancy like weighted sums or something. Thats what this was referring to. We can improve on it later

**Comment by etiennedi** [comment]
Closing as proposal accepted, currently being implemented in #976

**Comment by etiennedi** [comment]
Reopening as we realized, we need the distances per prop, not overall.

**Comment by etiennedi** [comment]
Resulting from a discussion on slack:

## Updated Proposal

## Changes since the original proposal:
* Each cross ref now has a `meta` object (visible only when `?meta=true` is set)
* The meta object contains the mean distance of the winning and losing group
* The overall meta object has no more `distance` field

## Notes / Limitations
* The overall meta object only shows the meta data of the last classification run. If multiple classifciations were run the meta data from previous requests is lost

## Example
Copied from updated source post
*Get with meta parameter `GET /v1/thing/<id>?meta=true`*
```
GET /v1/things/<id>
{
  "class": "Company",
  "id": "<id>",
  "schema": {
    "description": "Foo bar baz bazzar",
    "materialGroupOne": [
      {
         "beacon": "...",
         "meta": null  <----- indicates this has not been classified, but set by a user
      }
    ],
   "materialGroupTwo": [
      {
         "beacon": "...",
         "meta": {
            "classification": {
              "distanceWinning": 0.5,
              "distanceLosing": 0.6,
            },
          },
      }
    ]
  }
  "meta": {
    "classification": {
      "id": "<classificatinId>",
      "completed": <timestamp>,
      "classifiedFields": ["matericalGroupTwo"],
      "scope": ["materialGroupOne", "materialGroupTwo"]  <----- note that the scope was to classify both fields, but since the user had already set materialGroupOne the actual classified Fields (see one row up) is only a single field (materialGroupTwo)
      "basedOn": ["description"],
    }
  }
}
```

@fefi42 @bobvanluijt can you quickly confirm this proposal, then I can start implementing it. Thanks.

**Comment by bobvanluijt** [comment]
Confirmed, I will need a more detailed understanding of the changes for the docs but this isn't crucial for now.

**Comment by etiennedi** [comment]
Closing again as accepted, currently being implemented in #979.
